File: sdk/go/mics-hook-sdk/micshooksdk/mq_test.go
package micshooksdk

import "testing"

func TestMqTopicNames(t *testing.T) {
	if got := EventTopicName("t1"); got != "im-mics-t1-event" {
		t.Fatalf("unexpected event topic: %q", got)
	}
	if got := DlqTopicName("t1"); got != "im-mics-t1-event-dlq" {
		t.Fatalf("unexpected dlq topic: %q", got)
	}
}

func TestDecodeAndVerifyMqEvent_SignRequired(t *testing.T) {
	evt := MqEvent{
		TenantID:  "t1",
		EventType: EventSingleChatMsg,
		MsgID:     "m1",
		UserID:    "u1",
		DeviceID:  "d1",
		ToUserID:  "u2",
		EventData: []byte{1, 2, 3},
		Timestamp: 123,
		NodeID:    "node-1",
		TraceID:   "tr1",
	}

	signPayload := EncodeMqEvent(MqEvent{TenantID: evt.TenantID, EventType: evt.EventType, MsgID: evt.MsgID, UserID: evt.UserID, DeviceID: evt.DeviceID, ToUserID: evt.ToUserID, GroupID: evt.GroupID, EventData: evt.EventData, Timestamp: evt.Timestamp, NodeID: evt.NodeID, Sign: "", TraceID: evt.TraceID})
	evt.Sign = ComputeMqEventSignBase64("secret", signPayload)

	bytes := EncodeMqEvent(evt)
	res, err := DecodeAndVerifyMqEvent(bytes, "secret", true)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !res.Ok {
		t.Fatalf("expected ok=true, got reason=%q", res.Reason)
	}
	if res.Event.MsgID != "m1" {
		t.Fatalf("unexpected msg: %q", res.Event.MsgID)
	}
}

File: sdk/node/mics-hook-sdk/src/mq.test.ts
import { describe, expect, test } from "vitest";
import { dlqTopicName, eventTopicName, decodeMqEvent, isValidMqEventSign } from "./mq.js";
import { EventType, MqEventCodec } from "./proto/mics_hook.js";
import { computeMqEventSignBase64 } from "./signing.js";

describe("mq", () => {
  test("topic naming is per-tenant", () => {
    expect(eventTopicName("t1")).toBe("im-mics-t1-event");
    expect(dlqTopicName("t1")).toBe("im-mics-t1-event-dlq");
  });

  test("decodeMqEvent decodes protobuf bytes", () => {
    const evt = {
      tenantId: "t1",
      eventType: EventType.CONNECT_ONLINE,
      msgId: "",
      userId: "u1",
      deviceId: "d1",
      toUserId: "",
      groupId: "",
      eventData: new Uint8Array([1]),
      timestamp: 123,
      nodeId: "node-1",
      sign: "",
      traceId: "tr1",
    };
    const bytes = MqEventCodec.encode(evt).finish();
    const decoded = decodeMqEvent(bytes);
    expect(decoded.tenantId).toBe("t1");
    expect(decoded.userId).toBe("u1");
  });

  test("isValidMqEventSign verifies sign", () => {
    const evt = {
      tenantId: "t1",
      eventType: EventType.SINGLE_CHAT_MSG,
      msgId: "m1",
      userId: "u1",
      deviceId: "d1",
      toUserId: "u2",
      groupId: "",
      eventData: new Uint8Array([1]),
      timestamp: 123,
      nodeId: "node-1",
      sign: "",
      traceId: "tr1",
    };
    const sign = computeMqEventSignBase64("secret", { ...evt, sign: "ignored" });
    expect(isValidMqEventSign("secret", { ...evt, sign })).toBe(true);
    expect(isValidMqEventSign("secret", { ...evt, sign: sign + "x" })).toBe(false);
  });
});
